### ISSUE TITLE
`StreamWriter.DisposeAsync` disposes the underlying `Stream` synchronously

### ISSUE THREAD

**Issue by Neme12** [dotnet/runtime#88246]
### Description

I discovered issue this while filing #88244. By default, `StreamWriter.Dispose` also disposes the underlying `Stream`. `DisposeAsync` should dispose the underlying `Stream` asynchronously, but currently disposes it synchronously.

See
https://github.com/dotnet/runtime/blob/2a22adba0b94308a87c41323d26f915a82c74a18/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L241-L257
and
https://github.com/dotnet/runtime/blob/2a22adba0b94308a87c41323d26f915a82c74a18/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L211-L234

### Reproduction Steps

-

### Expected behavior

-

### Actual behavior

-

**Comment by Neme12** [comment]
I also think it should call `stream.Dispose()` as opposed to `stream.Close()`. If anyone creates their own stream and implements these methods differently, I think the expectation is that when you call `streamReader.Dispose()`, it would also call `Dispose()` on the stream. But that would be a breaking change of course.

**Comment by dersia** [comment]
I will take care of this as part of #88244

**Comment by Jozkee** [comment]
> I also think it should call stream.Dispose() instead of stream.Close()

I think for compatiblity reasons we should continue calling `Close()`.

**Comment by dersia** [comment]
> > I also think it should call stream.Dispose() instead of stream.Close()
> 
> I think for compatiblity reasons we should continue calling `Close()`.

yeah, `stream.dispose()` does not call `Close` on the stream, so that would be a behavioral change. I found some other issues on `StreamWriter.Dispose()`. Please see https://github.com/dotnet/runtime/issues/88244#issuecomment-1647543667

**Comment by Neme12** [comment]
> so that would be a behavioral change

Right. But fixing this issue (calling `CloseAsync` instead of `Close`) would be just as much of a behavioral change too.

**Comment by Neme12** [comment]
> yeah, `stream.dispose()` does not call `Close` on the stream, so that would be a behavioral change

Right. But fixing this issue would be a behavioral change too, so if it's OK to make this behavioral change, I think it should be OK to swap `Close` and `Dispose` too.

**Comment by Jozkee** [comment]
@Neme12 because changing `DisposeAsync` is not as impactful as changing `Dispose` which has been around for longer.

**Comment by NinoFloris** [comment]
We have an example in Npgsql https://github.com/npgsql/Npgsql/blob/main/src/Npgsql/NpgsqlRawCopyStream.cs#L353-L376 of a stream that should ideally take advantage of async if it's there.

NpgsqlRawCopyStream is the base stream used in our StreamReader/Writer derived types (same file, see NpgsqlCopyTextReader/Writer)

It seems its made very hard to have StreamWriter call DisposeAsync on the underlying stream, many methods do GetType() == typeof(StreamReader/Writer) checks and CloseStreamDuringDispose etc are all private, the * truly * async FlushAsync implementation is also private.

I would welcome any fixes or workarounds to get StreamReader/Writer to respect DisposeAsync.
